AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/SQLite/Statement+Model.swift

```diff
@@ -35,99 +33,159 @@ protocol StatementModelConvertible {
+            )
+            modelDictionary.updateValue(modelValue, forKeyPath: column)
+
+            // create lazy list for "many" relationships
+            if let id = modelValue as? String, field.isPrimaryKey {
+                let associations = schema.fields.values.filter {
+                    $0.isArray && $0.hasAssociation
+                }
+                let prefix = field.name.replacingOccurrences(of: field.name, with: "")
+                associations.forEach { association in
+                    let associatedField = association.associatedField?.name
+                    let lazyList = List<AnyModel>.lazyInit(associatedId: id,
+                                                           associatedWith: associatedField)
+                    let listKeyPath = prefix + association.name
+                    modelDictionary.updateValue(lazyList, forKeyPath: listKeyPath)
+                }
             }
         }
+        // swiftlint:disable:next force_cast
+        return modelDictionary as! ModelValues
+    }
 
-        if let id = values[schema.primaryKey.name] as? String {
-            // create instances of lazy-load lists of fields that represent "many" associations
-            schema.fields.values.filter { $0.isArray }.forEach { field in
-                // TODO extract this to a List utility
-                let lazyListDecodable: [String: Any?] = [
-                    "associatedId": id,
-                    "associatedField": field.associatedField?.name,
-                    "elements": []
-                ]
-                values[field.name] = lazyListDecodable
+}
+
+internal extension List {
+
+    /// Creates a data structure that is capable of initializing a `List<M>` with
+    /// lazy-load capabilities when the list is being decoded.
+    ///
+    /// See the `List.init(from:Decoder)` for details.
+    static func lazyInit(associatedId: String, associatedWith: String?) -> [String: Any?] {
+        return [
+            "associatedId": associatedId,
+            "associatedField": associatedWith,
+            "elements": []
+        ]
+    }
+}
+
+internal extension Dictionary where Key == String, Value == Any? {
+
+    /// Utility to create a `NSMutableDictionary` from a Swift `Dictionary<String, Any?>`.
+    func mutableCopy() -> NSMutableDictionary {
+        // swiftlint:disable:next force_cast
+        return (self as NSDictionary).mutableCopy() as! NSMutableDictionary
+    }
+}
+
+/// Extension that adds utilities to created nested values in a dictionary
+/// from a `keyPath` notation (e.g. `root.with.some.nested.prop`.
+internal extension NSMutableDictionary {
+
+    /// Utility to allows Swift standard types to be used in `setObject`
+    /// of the `NSMutableDictionary`.
+    ///
+    /// - Parameters:
+    ///   - value: the value to be set
+    ///   - key: the key as a `String`
+    func updateValue(_ value: Value?, forKey key: String) {
+        setObject(value as Any, forKey: key as NSString)
```

Comment:
If value is nil,  i'm not sure if this works, because I don't think NSDictionaries can hold `nil` (i believe in obj-c we'd have to set it to `[NSNull null]`?

Reply:
Good call, fixing it!